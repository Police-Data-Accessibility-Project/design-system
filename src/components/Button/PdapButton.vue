<template>
	<button :class="classes">
		<slot />
	</button>
</template>

<script setup lang="ts">
// Imports
import { reactive } from 'vue';

// Types
import { PdapButtonProps } from './types';

// Props
const props = withDefaults(defineProps<PdapButtonProps>(), {
	intent: 'primary',
	isLoading: false,
});

// CSS class map
const classes = reactive({
	'pdap-button': true,
	[`pdap-button-${props.intent}`]: Boolean(props.intent),
	'pdap-button-loading': props.isLoading,
});
</script>

<script lang="ts">
/**
 * # `Button`
 * Styled button with PDAP branding. DOM attributes are forwarded to the HTML element, and most functionality works well this way.
 *
 * ## Props
 * @prop {'primary' | 'secondary' | 'tertiary'} intent controls the styling of the button
 * @prop {boolean} isLoading loading state of button
 *
 */
export default {
	name: 'PdapButton',
};
</script>

<style>
@tailwind components;

@layer components {
	.pdap-button {
		@apply cursor-pointer border-2 border-brand-gold decoration-0 disabled:opacity-50 font-semibold inline-block px-6 py-2 rounded-none text-center text-lg w-full;
		@apply hover:brightness-85 lg:text-xl sm:max-w-max;
	}

	.pdap-button-primary {
		@apply pdap-button bg-brand-gold text-neutral-50;
	}

	.pdap-button-primary[type='submit'] {
		@apply bg-brand-gold;
	}

	.pdap-button-secondary {
		@apply pdap-button bg-transparent border-2 border-brand-gold text-brand-gold;
	}

	.pdap-button-secondary[type='submit'] {
		@apply bg-transparent;
	}

	.pdap-button-tertiary {
		@apply pdap-button bg-transparent border-2 border-transparent text-neutral-950;
	}

	.pdap-button-tertiary[type='submit'] {
		@apply bg-transparent;
	}
}
</style>
