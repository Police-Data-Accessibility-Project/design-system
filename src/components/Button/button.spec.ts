// Component
import PdapButton from './PdapButton.vue';

// Utils
import { mount } from '@vue/test-utils';
import { describe, expect, test } from 'vitest';

// Test
describe('Renders button component', () => {
	// Render
	test('Renders a button', () => {
		const wrapper = mount(PdapButton, {
			slots: {
				default: 'Button Content',
			},
		});

		expect(wrapper.find('button').exists()).toBe(true);
		expect(wrapper.classes()).toContain('pdap-button');
		expect(wrapper.html()).toContain('Button Content');
		expect(wrapper.html()).toMatchSnapshot();
	});

	// Props
	// Props - intent
	test('Renders primary button', () => {
		const wrapper = mount(PdapButton, { props: { intent: 'primary' } });
		expect(wrapper.props()).toStrictEqual({
			intent: 'primary',
			isLoading: false,
		});
		expect(wrapper.classes()).toContain('pdap-button-primary');
	});

	test('Renders secondary button', () => {
		const wrapper = mount(PdapButton, { props: { intent: 'secondary' } });
		expect(wrapper.props().intent).toBe('secondary');
		expect(wrapper.classes()).toContain('pdap-button-secondary');
	});

	// Props - isLoading
	test('Renders loading button', () => {
		const wrapper = mount(PdapButton, { props: { isLoading: true } });
		expect(wrapper.props().isLoading).toBe(true);
		expect(wrapper.classes()).toContain('pdap-button-loading');
	});

	// Events
	test('Handles a click event', () => {
		const wrapper = mount(PdapButton);
		wrapper.find('button').trigger('click');
	});
	test('Handles a keydown event', () => {
		const wrapper = mount(PdapButton);
		wrapper.find('button').trigger('keydown');
	});
});
