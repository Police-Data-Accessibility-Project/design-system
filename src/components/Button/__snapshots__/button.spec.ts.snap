// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders button component > Renders a button 1`] = `
<button class="pdap-button pdap-button-primary">
  <transition-group-stub appear="false" css="true" persisted="false">Button Content
    <!--v-if-->
    <!--v-if-->
  </transition-group-stub>
</button>
`;
